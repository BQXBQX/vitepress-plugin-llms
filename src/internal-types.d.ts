import type { GrayMatterFile } from 'gray-matter'
import type { ResolvedConfig } from 'vite'
import type { SiteConfig, UserConfig } from 'vitepress'

/** Represents a prepared file, including its title and path. */
export type PreparedFile = {
	/**
	 * The title of the file.
	 *
	 * @example 'Guide'
	 */
	title: string

	/**
	 * The absolute path to the file.
	 *
	 * @example 'guide/getting-started.md'
	 */
	path: string

	/**
	 * The prepared file itself.
	 *
	 * @example
	 * ```typescript
	 * {
	 *   data: {
	 *      title: 'Guide'
	 *   },
	 *   content: 'Content goes here'
	 *   orig: '---\ntitle: Guide\n---\n\nContent goes here'
	 * }
	 * ```
	 */
	file: GrayMatterFile<Input>
}

export interface VitePressConfig extends Omit<UserConfig, keyof ResolvedConfig>, ResolvedConfig {
	vitepress: SiteConfig
}

/** Represents the link extension options for generated links. */
export type LinksExtension = string | '.md' | '.html'

export type NotUndefined<T> = {
	[K in keyof T]-?: Exclude<T[K], undefined>
}
