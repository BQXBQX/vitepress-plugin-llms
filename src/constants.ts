/** Default template for the `llms.txt` file. */
export const defaultLLMsTxtTemplate = `\
# {title}

{description}

## Table of Contents

{toc}`
