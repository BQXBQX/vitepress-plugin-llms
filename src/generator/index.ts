export * from './llms-full-txt'
export * from './llms-txt'
export * from './page-generator'
export * from './toc'
