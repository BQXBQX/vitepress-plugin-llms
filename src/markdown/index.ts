export * from './markdown-it-plugins'
export * from './remark-plugins'
