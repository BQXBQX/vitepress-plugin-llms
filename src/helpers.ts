import fs from 'node:fs'
import path from 'node:path'
import type { LlmstxtSettings, PreparedFile, VitePressConfig } from './types'
import matter, { type GrayMatterFile, type Input } from 'gray-matter'
// @ts-ignore
import markdownTitle from 'markdown-title'
import { stripHtml } from 'string-strip-html'
import { defaultLLMsTxtTemplate } from './constants'

/**
 * Splits a file path into its directory and file components.
 *
 * @param filepath - The path to the file.
 * @returns An object containing the directory and file name.
 */
export const splitDirAndFile = (filepath: string) => ({
	dir: path.dirname(filepath),
	file: path.basename(filepath),
})

/**
 * Strips the file extension from a given file path.
 *
 * @param filepath - The path to the file.
 * @returns The filename without the extension.
 */
export const stripExt = (filepath: string) => {
	const { dir, file } = splitDirAndFile(filepath)

	return path.join(dir, path.basename(file, path.extname(file)))
}

/**
 * Strips the file extension from a given file path using POSIX format.
 *
 * @param filepath - The path to the file.
 * @returns The filename without the extension in POSIX format.
 */
export const stripExtPosix = (filepath: string) => {
	const { dir, file } = splitDirAndFile(filepath)

	return path.posix.join(dir, path.basename(file, path.extname(file)))
}

/**
 * Extracts the title from a markdown file.
 *
 * @param content - The content of the markdown file.
 * @param vitepressConfig - The VitePress configuration.
 * @returns The title of the markdown file, or a default title if none is found.
 */
export function extractTitle(
	file: GrayMatterFile<Input>,
	vitepressConfig: VitePressConfig,
): string {
	return (
		file.data?.title ||
		vitepressConfig?.vitepress?.userConfig?.title ||
		file.data?.titleTemplate ||
		vitepressConfig?.vitepress?.userConfig?.titleTemplate ||
		markdownTitle(
			stripHtml(file.content, { stripTogetherWithTheirContents: ['*'] }).result,
		)
	)
}

/**
 * Generates a Table of Contents (TOC) for the provided prepared files.
 *
 * Each entry in the TOC is formatted as a markdown link to the corresponding
 * text file.
 *
 * @param preparedFiles - An array of prepared files.
 * @param srcDir - The VitePress source directory.
 * @returns A string representing the formatted Table of Contents.
 */
export function generateTOC(
	preparedFiles: PreparedFile[],
	srcDir: VitePressConfig['vitepress']['srcDir'],
) {
	let tableOfContent = ''

	for (const file of preparedFiles) {
		const relativePath = path.relative(srcDir, file.path)
		tableOfContent += `- [${file.title}](/${stripExtPosix(relativePath)}.md)\n`
	}

	return tableOfContent
}

/**
 * Creates a regular expression to match a specific template variable in the format `{key}`.
 *
 * @param key - The name of the template variable to match.
 * @returns A case-insensitive regular expression that detects `{key}` occurrences in a string.
 *
 * @example
 * ```ts
 * const regex = templateVariable('NAME');
 * console.log(regex.test('Hello {NAME}')); // true
 * ```
 */
const templateVariable = (key: string) =>
	new RegExp(`(\n\s*\n)?\{${key}\}`, 'gi')

/**
 * Replaces occurrences of a template variable `{variable}` in a given content string with a provided value.
 * If the value is empty or undefined, it falls back to a specified fallback value.
 *
 * @param content - The template string containing placeholders.
 * @param variable - The template variable name to replace.
 * @param value - The value to replace the variable with.
 * @param fallback - An optional fallback value if `value` is empty.
 * @returns A new string with the template variable replaced.
 *
 * @example
 * ```ts
 * const template = 'Hello {name}!';
 * const result = replaceTemplateVariable(template, 'name', 'Alice', 'User');
 * console.log(result); // 'Hello Alice!'
 * ```
 */
export function replaceTemplateVariable(
	content: string,
	variable: string,
	value: string,
	fallback?: string,
) {
	return content.replace(templateVariable(variable), (_, prefix) => {
		const val = value?.length ? value : fallback?.length ? fallback : ''
		return val ? `${prefix ? '\n\n' : ''}${val}` : ''
	})
}

/**
 * Expands a template string by replacing multiple template variables with their corresponding values.
 *
 * @param template - The template string containing placeholders.
 * @param values - An object mapping variable names to their respective values.
 * @returns A string with all template variables replaced.
 *
 * @example
 * ```ts
 * const template = 'Hello {name}, welcome to {place}!';
 * const values = { name: 'Alice', place: 'Wonderland' };
 * const result = expandTemplate(template, values);
 * console.log(result); // 'Hello Alice, welcome to Wonderland!'
 * ```
 */
export function expandTemplate(
	template: string,
	values: { [key: string]: string },
): string {
	return Object.entries(values).reduce(
		(result, [key, value]) => replaceTemplateVariable(result, key, value),
		template,
	)
}

/**
 * Generates a LLMs.txt file with a table of contents and links to all documentation sections.
 *
 * @param preparedFiles - An array of prepared files.
 * @param indexMd - Path to the main documentation file `index.md`.
 * @param vitepressConfig - The VitePress configuration.
 * @param customLLMsTxtTemplate - Template to use for generating `llms.txt`.
 * @param customTemplateVariables - Custom variables for `customLLMsTxtTemplate`.
 * @returns A string representing the content of the `llms.txt` file.
 *
 * @example
 * ```markdown
 * # Shadcn for Vue
 *
 * > Beautifully designed components built with Radix Vue and Tailwind CSS.
 *
 * ## Table of Contents
 *
 * - [Getting started](/docs/getting-started.md)
 * - [About](/docs/about.md)
 * - ...
 * ```
 *
 * @see https://llmstxt.org/#format
 */
export function generateLLMsTxt(
	preparedFiles: PreparedFile[],
	indexMd: string,
	vitepressConfig: VitePressConfig,
	srcDir: VitePressConfig['vitepress']['srcDir'],
	customLLMsTxtTemplate: LlmstxtSettings['customLLMsTxtTemplate'] = defaultLLMsTxtTemplate,
	customTemplateVariables: LlmstxtSettings['customTemplateVariables'] = {},
) {
	// @ts-expect-error
	matter.clearCache()
	const indexMdFile = matter(fs.readFileSync(indexMd, 'utf-8') as string)

	// biome-ignore lint/suspicious/noExplicitAny:
	const defaults: Record<string, any> = {}

	if (!customTemplateVariables.title) {
		defaults.title =
			indexMdFile.data?.hero?.name ||
			extractTitle(indexMdFile, vitepressConfig) ||
			'LLMs Documentation'
	}

	if (!customTemplateVariables.description) {
		defaults.description =
			indexMdFile.data?.hero?.text ||
			vitepressConfig?.vitepress?.userConfig?.description ||
			indexMdFile?.data?.description ||
			indexMdFile.data?.titleTemplate ||
			'This file contains links to all documentation sections.'
	}

	if (!customTemplateVariables.details) {
		defaults.details =
			indexMdFile.data?.hero?.tagline || indexMdFile.data?.tagline
	}

	if (!customTemplateVariables.toc) {
		defaults.toc = generateTOC(preparedFiles, srcDir)
	}

	return expandTemplate(customLLMsTxtTemplate, {
		...defaults,
		...customTemplateVariables,
	})
}

/**
 * Generates a `llms-full.txt` file content with all documentation in one file.
 *
 * @param preparedFiles - An array of prepared files.
 * @param srcDir - The source directory for the files.
 * @returns A string representing the full content of the LLMs.txt file.
 */
export function generateLLMsFullTxt(
	preparedFiles: PreparedFile[],
	srcDir: VitePressConfig['vitepress']['srcDir'],
) {
	const llmsFullTxtContent = preparedFiles
		.map((file) => {
			const relativePath = path.relative(srcDir, file.path)

			file.file.data = {
				url: `/${stripExtPosix(relativePath)}.md`,
			}

			return matter.stringify(file.file.content, file.file.data)
		})
		.join('\n---\n\n')

	return llmsFullTxtContent
}
