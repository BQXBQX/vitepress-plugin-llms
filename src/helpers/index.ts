import fs from 'node:fs/promises'
import path from 'node:path'

import matter from 'gray-matter'

import type { DefaultTheme } from 'vitepress'
import { defaultLLMsTxtTemplate } from '../constants'
import type { LlmstxtSettings, PreparedFile, VitePressConfig } from '../types'
import { generateTOC } from './toc'
import { expandTemplate, extractTitle, generateMetadata } from './utils'

/**
 * Generates a LLMs.txt file with a table of contents and links to all documentation sections.
 *
 * @param preparedFiles - An array of prepared files.
 * @param indexMd - Path to the main documentation file `index.md`.
 * @param vitepressConfig - The VitePress configuration.
 * @param LLMsTxtTemplate - Template to use for generating `llms.txt`.
 * @param templateVariables - Template variables for `customLLMsTxtTemplate`.
 * @param domain - The base domain for the generated links.
 * @param sidebar - Optional sidebar configuration for organizing the TOC.
 * @returns A string representing the content of the `llms.txt` file.
 *
 * @example
 * ```markdown
 * # Shadcn for Vue
 *
 * > Beautifully designed components built with Radix Vue and Tailwind CSS.
 *
 * ## Table of Contents
 *
 * - [Getting started](/docs/getting-started.md)
 * - [About](/docs/about.md)
 * - ...
 * ```
 *
 * @see https://llmstxt.org/#format
 */
export async function generateLLMsTxt(
	preparedFiles: PreparedFile[],
	indexMd: string,
	srcDir: VitePressConfig['vitepress']['srcDir'],
	LLMsTxtTemplate: LlmstxtSettings['customLLMsTxtTemplate'] = defaultLLMsTxtTemplate,
	templateVariables: LlmstxtSettings['customTemplateVariables'] = {},
	vitepressConfig?: VitePressConfig['vitepress']['userConfig'],
	domain?: LlmstxtSettings['domain'],
	sidebar?: DefaultTheme.Sidebar,
): Promise<string> {
	// @ts-expect-error
	matter.clearCache()
	const indexMdContent = await fs.readFile(indexMd, 'utf-8')
	const indexMdFile = matter(indexMdContent as string)

	templateVariables.title ??=
		indexMdFile.data?.hero?.name ||
		indexMdFile.data?.title ||
		vitepressConfig?.title ||
		vitepressConfig?.titleTemplate ||
		extractTitle(indexMdFile) ||
		'LLMs Documentation'

	templateVariables.description ??=
		indexMdFile.data?.hero?.text ||
		vitepressConfig?.description ||
		indexMdFile?.data?.description ||
		indexMdFile.data?.titleTemplate

	templateVariables.details ??=
		indexMdFile.data?.hero?.tagline ||
		indexMdFile.data?.tagline ||
		(!templateVariables.description &&
			'This file contains links to all documentation sections.')

	templateVariables.toc ??= await generateTOC(
		preparedFiles,
		srcDir,
		domain,
		sidebar || vitepressConfig?.themeConfig?.sidebar,
	)

	return expandTemplate(LLMsTxtTemplate, templateVariables)
}

/**
 * Generates a `llms-full.txt` file content with all documentation in one file.
 *
 * @param preparedFiles - An array of prepared files.
 * @param srcDir - The source directory for the files.
 * @param domain - The base domain for the generated links.
 * @returns A string representing the full content of the LLMs.txt file.
 */
export function generateLLMsFullTxt(
	preparedFiles: PreparedFile[],
	srcDir: VitePressConfig['vitepress']['srcDir'],
	domain?: LlmstxtSettings['domain'],
) {
	const llmsFullTxtContent = preparedFiles
		.map((preparedFile) => {
			const relativePath = path.relative(srcDir, preparedFile.path)

			return matter.stringify(
				preparedFile.file.content,
				generateMetadata(preparedFile.file, domain, relativePath),
			)
		})
		.join('\n---\n\n')

	return llmsFullTxtContent
}
