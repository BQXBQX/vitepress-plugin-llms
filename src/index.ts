import type { Plugin } from 'vitepress'
import type { ViteDevServer } from 'vite'

import fs from 'node:fs'
import path from 'node:path'

import pc from 'picocolors'
import { minimatch } from 'minimatch'
import {
	extractTitle,
	generateLLMsFullTxt,
	generateLLMsTxt,
	stripExt,
	stripExtPosix,
} from './helpers'
import log from './logger'
import type { LlmstxtSettings, PreparedFile, VitePressConfig } from './types'
import matter from 'gray-matter'
import { defaultLLMsTxtTemplate } from './constants'
import { name as packageName } from '../package.json'

const PLUGIN_NAME = packageName

/**
 * [VitePress](http://vitepress.dev/) plugin for generating raw documentation
 * for **LLMs** in Markdown format which is much lighter and more efficient for LLMs
 *
 * @param [userSettings={}] - Plugin settings.
 *
 * @see https://github.com/okineadev/vitepress-plugin-llms
 * @see https://llmstxt.org/
 */
export default function llmstxt(userSettings: LlmstxtSettings = {}): Plugin {
	// Create a settings object with defaults explicitly merged
	const settings: LlmstxtSettings = {
		generateLLMsFullTxt: true,
		generateLLMsTxt: true,
		ignoreFiles: [],
		customLLMsTxtTemplate: defaultLLMsTxtTemplate,
		...userSettings,
	}

	// Store the resolved Vite config
	let config: VitePressConfig

	// Set to store all markdown file paths
	const mdFiles: Set<string> = new Set()

	// Flag to identify which build we're in
	let isSsrBuild = false

	return {
		name: PLUGIN_NAME,
		enforce: 'post', // Run after other plugins

		/** Resolves the Vite configuration and sets up the working directory. */
		configResolved(resolvedConfig) {
			config = resolvedConfig as VitePressConfig
			if (settings.workDir) {
				settings.workDir = path.resolve(
					config.vitepress.srcDir,
					settings.workDir as string,
				)
			} else {
				settings.workDir = config.vitepress.srcDir
			}
			// Detect if this is the SSR build
			isSsrBuild = !!resolvedConfig.build?.ssr
			log.info(
				`Plugin initialized ${isSsrBuild ? pc.dim('(SSR build)') : pc.dim('(client build)')}`,
			)
		},

		/** Configures the development server to handle `llms.txt` and markdown files for LLMs. */
		configureServer(server: ViteDevServer) {
			log.info('Development server configured to serve markdown files')
			server.middlewares.use((req, res, next) => {
				if (req.url?.endsWith('.md') || req.url?.endsWith('.txt')) {
					try {
						// Try to read and serve the markdown file
						const filePath = path.resolve(
							config.vitepress?.outDir ?? 'dist',
							`${stripExt(req.url)}.md`,
						)
						const content = fs.readFileSync(filePath, 'utf-8')
						res.setHeader('Content-Type', 'text/markdown')
						res.end(content)
						return
					} catch (e) {
						// If file doesn't exist or can't be read, continue to next middleware
						next()
					}
				}

				// Pass to next middleware if not handled
				next()
			})
		},

		/**
		 * Resets the collection of markdown files when the build starts.
		 * This ensures we don't include stale data from previous builds.
		 */
		buildStart() {
			mdFiles.clear()
			log.info('Build started, file collection cleared')
		},

		/**
		 * Processes each file that Vite transforms and collects markdown files.
		 * @param _ - The file content (not used).
		 * @param id - The file identifier (path).
		 * @returns null if the file is processed, otherwise returns the original content.
		 */
		transform(_, id: string) {
			if (!id.endsWith('.md')) {
				return null
			}

			// Skip files outside workDir if it's configured
			if (!id.startsWith(settings.workDir as string)) {
				return null
			}

			if (settings.ignoreFiles?.length) {
				for (const pattern of settings.ignoreFiles) {
					if (
						typeof pattern === 'string' &&
						minimatch(path.relative(settings.workDir as string, id), pattern)
					) {
						return null
					}
				}
			}

			// Add markdown file path to our collection
			mdFiles.add(id)
			// Return null to avoid modifying the file
			return null
		},

		/**
		 * Runs only in the client build (not SSR) after completion.
		 * This ensures the processing happens exactly once.
		 */
		generateBundle() {
			// Skip processing during SSR build
			if (isSsrBuild) {
				log.info('Skipping file generation in SSR build')
				return
			}

			const outDir = config.vitepress?.outDir ?? 'dist'

			// Create output directory if it doesn't exist
			if (!fs.existsSync(outDir)) {
				log.info(`Creating output directory: ${pc.cyan(outDir)}`)
				fs.mkdirSync(outDir, { recursive: true })
			}

			const mdFilesList = Array.from(mdFiles)
			const fileCount = mdFilesList.length

			// Skip if no files found
			if (fileCount === 0) {
				log.warn('No markdown files found to process')
				return
			}

			log.info(`Processing ${pc.bold(fileCount.toString())} markdown files...`)

			const preparedFiles: PreparedFile[] = []

			// Copy all markdown files to output directory
			for (const file of mdFilesList) {
				const relativePath = path.relative(settings.workDir as string, file)
				const targetPath = path.resolve(outDir, relativePath)

				try {
					const fileContent = matter(fs.readFileSync(file, 'utf-8'))
					const title = extractTitle(fileContent.orig.toString()) || 'Untitled'

					preparedFiles.push({ title, path: file })

					// Ensure target directory exists
					fs.mkdirSync(path.dirname(targetPath), { recursive: true })

					fileContent.data = {
						url: `/${stripExtPosix(relativePath)}.md`,
					}

					if (fileContent.data?.description?.length) {
						// biome-ignore lint/correctness/noSelfAssign: <explanation>
						fileContent.data.description = fileContent.data?.description
					}

					// Copy file to output directory
					fs.writeFileSync(
						targetPath,
						matter.stringify(fileContent.content, fileContent.data),
					)
					log.success(`Copied ${pc.cyan(relativePath)} to output directory`)
				} catch (error) {
					// @ts-ignore
					log.error(`Failed to copy ${pc.cyan(relativePath)}: ${error.message}`)
				}
			}

			// Sort files by title for better organization
			preparedFiles.sort((a, b) => a.title.localeCompare(b.title))

			// Generate llms.txt - table of contents with links
			if (settings.generateLLMsTxt) {
				const llmsTxtPath = path.resolve(outDir, 'llms.txt')

				const llmsTxt = generateLLMsTxt(
					preparedFiles,
					path.resolve(settings.workDir as string, 'index.md'),
					settings.workDir as string,
					settings.customLLMsTxtTemplate,
					settings.customTemplateVariables,
				)

				fs.writeFileSync(llmsTxtPath, llmsTxt, 'utf-8')
				log.success(
					`Generated ${pc.cyan('llms.txt')} with ${pc.bold(fileCount.toString())} documentation sections`,
				)
			}

			// Generate llms-full.txt - all content in one file
			if (settings.generateLLMsFullTxt) {
				const llmsFullTxtPath = path.resolve(outDir, 'llms-full.txt')

				log.info(`Generating ${pc.cyan('llms-full.txt')}...`)

				const llmsFullTxt = generateLLMsFullTxt(
					preparedFiles,
					settings.workDir as string,
				)

				// Write content to llms-full.txt
				fs.writeFileSync(llmsFullTxtPath, llmsFullTxt, 'utf-8')
				log.success(
					`Generated ${pc.cyan('llms-full.txt')} with ${pc.bold(fileCount.toString())} markdown files`,
				)
			}
		},
	}
}
