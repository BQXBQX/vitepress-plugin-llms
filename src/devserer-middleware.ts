import fs from 'node:fs/promises'
import path from 'node:path'
import pc from 'picocolors'
import type { Connect } from 'vite'
import log from './helpers/logger'
import { stripExt } from './helpers/utils'
import type { VitePressConfig } from './types'

export default (vitePressConfig: VitePressConfig['vitepress']) =>
	(async (req, res, next): Promise<void> => {
		if (req.url?.endsWith('.md') || req.url?.endsWith('.txt')) {
			try {
				// Try to read and serve the markdown file
				const filePath = path.resolve(vitePressConfig?.outDir ?? 'dist', `${stripExt(req.url)}.md`)
				const content = await fs.readFile(filePath, 'utf-8')
				res.setHeader('Content-Type', 'text/plain; charset=utf-8')
				res.end(content)
				return
			} catch (e) {
				// If file doesn't exist or can't be read, continue to next middleware
				log.warn(`Failed to return ${pc.cyan(req.url)}: File not found`)
				next()
			}
		}

		// Pass to next middleware if not handled
		next()
	}) satisfies Connect.NextHandleFunction
