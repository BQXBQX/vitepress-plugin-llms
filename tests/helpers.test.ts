import { describe, expect, it, mock, test } from 'bun:test'
import {
	fakeCustomLlmsTxtTemplate,
	fakeIndexMd,
	fakeMarkdownDocument,
} from './resources'

const srcDir = 'docs'

const readFileSync = mock((path) =>
	path === `${srcDir}/index.md` ? fakeIndexMd : fakeMarkdownDocument,
)

mock.module('node:fs', () => ({
	default: { readFileSync },
	readFileSync,
}))

import {
	generateLLMsFullTxt,
	generateLLMsTxt,
	generateTOC,
	// @ts-ignore
} from '../src/helpers'

import type { PreparedFile, VitePressConfig } from '../src/types'
// @ts-ignore
import { defaultLLMsTxtTemplate } from '../src/constants'

const preparedFilesSample: PreparedFile[] = [
	{
		title: 'My Title',
		path: `${srcDir}/index.md`,
	},
	{
		title: 'My Title 2',
		path: `${srcDir}/test/test.md`,
	},
]

describe('generateTOC', () => {
	it('generates a table of contents', () => {
		expect(generateTOC(preparedFilesSample, srcDir)).toBe(
			'- [My Title](/index.md)\n- [My Title 2](/test/test.md)\n',
		)
	})
})

describe('generateLLMsTxt', () => {
	it('generates a `llms.txt` file', () => {
		expect(
			generateLLMsTxt(
				preparedFilesSample,
				`${srcDir}/index.md`,
				{} as VitePressConfig,
				srcDir,
				defaultLLMsTxtTemplate,
			),
		).toMatchSnapshot()
	})
	it('works correctly with a custom template', () => {
		expect(
			generateLLMsTxt(
				preparedFilesSample,
				`${srcDir}/index.md`,
				{} as VitePressConfig,
				srcDir,
				fakeCustomLlmsTxtTemplate,
			),
		).toMatchSnapshot()
	})
	it('works correctly with a custom template variables', () => {
		expect(
			generateLLMsTxt(
				preparedFilesSample,
				`${srcDir}/index.md`,
				{} as VitePressConfig,
				srcDir,
				defaultLLMsTxtTemplate,
				{ title: 'foo', description: 'bar', toc: 'zoo' },
			),
		).toMatchSnapshot()
	})

	it('works correctly with a custom template and variables', () => {
		expect(
			generateLLMsTxt(
				preparedFilesSample,
				`${srcDir}/index.md`,
				{} as VitePressConfig,
				srcDir,
				'# {foo}\n\n**{bar}**\n\n{zoo}',
				{ title: 'foo', description: 'bar', toc: 'zoo' },
			),
		).toMatchSnapshot()
	})
})

describe('generateLLMsFullTxt', () => {
	it('generates a `llms-full.txt` file', () => {
		expect(generateLLMsFullTxt(preparedFilesSample, srcDir)).toMatchSnapshot()
	})
})
