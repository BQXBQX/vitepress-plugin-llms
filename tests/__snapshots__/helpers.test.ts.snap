// Bun Snapshot v1, https://goo.gl/fbAQLP

exports[`generateLLMsTxt generates a \`llms.txt\` file 1`] = `
"# My Site

> This file contains links to all documentation sections.

## Table of Contents

- [My Title](/index.md)
- [My Title 2](/test/test.md)
"
`;

exports[`generateLLMsTxt works correctly with a custom template 1`] = `
"# Custom title

> Custom description

## TOC

- [My Title](/index.md)
- [My Title 2](/test/test.md)
"
`;

exports[`generateLLMsTxt works correctly with a custom template variables 1`] = `
"# foo

> bar

## Table of Contents

zoo"
`;

exports[`generateLLMsTxt works correctly with a custom template and variables 1`] = `
"# {foo}

**{bar}**

{zoo}"
`;

exports[`generateLLMsFullTxt generates a \`llms-full.txt\` file 1`] = `
"---
url: /index.md
---
# Welcome
Content goes here

---

---
url: /test/test.md
---
# Some cool stuff
"
`;
